src/api/githubService.js

```diff
@@ -0,0 +1,286 @@
+  }
+
+  return return_data;
+};
+
+const _getWorkflowCheckRuns = async (checkSuiteId) => {
+  const workflowCheckRunsApi = `https://api.github.com/repos/mozilla-mobile/mozilla-vpn-client/check-suites/${checkSuiteId}/check-runs`;
+  let return_data = [];
+
+  try {
+    const {
+      data: { check_runs },
+    } = await axios.get(workflowCheckRunsApi);
+    for (let check_run of check_runs) {
+      if (check_run.name !== "Build Test Client") {
+        const run = {
+          id: check_run.id,
+          name: check_run.name,
+          status: check_run.status,
+          conclusion: check_run.conclusion,
+          head_sha: check_run.head_sha,
+          started: check_run.started_at,
+          completed: check_run.completed_at,
+          annotations: check_run.output.annotations_count,
+        };
+
+        return_data.push(run);
+      }
+    }
+  } catch (err) {
+    console.log("Unable to fetch _getWorkflowCheckRuns", err);
+    return;
+  }
+
+  return return_data;
+};
+
+const _getWorkflowCheckRunAnnotation = async (checkRunId) => {
+  const workflowCheckRunAnnotationApi = `https://api.github.com/repos/mozilla-mobile/mozilla-vpn-client/check-runs/${checkRunId}/annotations`;
+  let return_data = [];
+
+  try {
+    const { data } = await axios.get(workflowCheckRunAnnotationApi);
+    for (let run_annotation of data) {
+      const annotation = {
+        message: run_annotation.message,
+        annotation_level: run_annotation.annotation_level,
+        start_line: run_annotation.start_line,
+        end_line: run_annotation.end_line,
+      };
+
+      return_data.push(annotation);
+    }
+  } catch (err) {
+    console.log("Unable to fetch _getWorkflowCheckRunAnnotation", err);
+    return;
+  }
+
+  return return_data;
+};
+
+/// full api
+export const getFullTestWorkflowReport = async () => {
+  let return_data = {
+    fetched: moment().format().valueOf(),
+    workflow_runs_data: [],
+  };
+
+  if (await _getLocalData()) {
+    const localData = await _getDataFromLocal();
+    return_data.fetch_message =
+      "Its too early to request again, clear expiretime";
+    return_data.workflow_runs_data = localData.workflow_runs_data;
+    return return_data;
+  }
+
+  if (await _checkForUpdate()) {
+    const localData = await _getDataFromLocal();
+    return_data.fetch_message = "No New Workflows Available";
+    return_data.workflow_runs_data = localData.workflow_runs_data;
+    return return_data;
+  }
+
+  try {
+    // get latest workflow runs for main branch
+    const workflows = await _getSingleWorkflow(functionalTestsWorkflowId);
+
+    // get checks (functional test matrix run) for each workflow run
+    for (let i = 0; i < workflows.length; i++) {
+      let workflow = {
+        workflow_run: {},
+        test_runs: [],
+        passed_count: 0,
+        failed_count: 0,
+        flaked_count: 0,
+      };
+
+      workflow.workflow_run = workflows[i];
+      const delayTime = Math.floor(Math.random() * 1000) + 200;
```

Comment:
why this delayTime?

Reply:
rate limiting, wanted to add some sort of randomization to the requests.